packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1025 @@
+                this.focusState.focusType = TableFocusType.row;
+                return this.scrollToAndFocusRow(0);
+            }
+            case TableFocusType.columnHeader: {
+                this.focusState.focusType = TableFocusType.cell;
+                return this.scrollToAndFocusRow(0);
+            }
+            case TableFocusType.row:
+            case TableFocusType.rowSelectionCheckbox:
+            case TableFocusType.cell:
+                if (
+                    this.focusState.rowIndex! + 1
+                    < this.table.tableData.length
+                ) {
+                    return this.scrollToAndFocusRow(
+                        this.focusState.rowIndex! + 1
+                    );
+                }
+                return false;
+            default:
+                break;
+        }
+
+        return false;
+    }
+
+    private onTabPressed(shiftKeyPressed: boolean): boolean {
+        const activeElement = this.getActiveElement();
+        if (activeElement === null || activeElement === this.table) {
+            return false;
+        }
+
+        let focusableElement: HTMLElement | undefined;
+        if (this.hasRowOrCellFocusType()) {
+            const row = this.getCurrentRow()!;
+            const focusableRowElements = row.getFocusableElements();
+            if (!shiftKeyPressed) {
+                let cellIndex = -1;
+                switch (this.focusState.focusType) {
+                    case TableFocusType.row:
+                        if (focusableRowElements.selectionCheckbox) {
+                            focusableElement = focusableRowElements.selectionCheckbox;
+                            this.focusState.focusType = TableFocusType.rowSelectionCheckbox;
+                        } else {
+                            cellIndex = 0;
+                        }
+                        break;
+                    case TableFocusType.rowSelectionCheckbox:
+                        cellIndex = 0;
+                        break;
+                    case TableFocusType.cell:
+                    case TableFocusType.cellActionMenu:
+                    case TableFocusType.cellContent:
+                        cellIndex = this.focusState.columnIndex!;
+                        if (
+                            this.focusState.focusType
+                            === TableFocusType.cellActionMenu
+                        ) {
+                            cellIndex += 1;
+                        }
+                        break;
+                    default:
+                        break;
+                }
+                let cellContentIndex = 0;
+                if (this.focusState.focusType === TableFocusType.cellContent) {
+                    cellContentIndex = this.focusState.cellContentIndex! + 1;
+                }
+                while (
+                    cellIndex >= 0
+                    && cellIndex < this.table.visibleColumns.length
```

Comment:
Maybe we can make a couple of private helper methods to help with code readability (reduce function length). Somethine like `setNextTabFocusState(cellIndex: number)` and `setPreviousTabFocusState(...)`? Possibly even `getNextTabCellIndex` and `getPreviousTabCellIndex` that just encapsulates the switch logic.

Reply:
Agreed. I'll want to do another pass on code sharing/ helper methods if people are otherwise satisfied with the general approach on tracking focus state that I have here. I wanted to see if there were any large course corrections needed first though.

Reply:
I'm ok with the overall architecture of managing the state from this class. For me the state machine readability is the biggest course correction. I left some brainstorm-y ideas on that in this thread: https://github.com/ni/nimble/pull/2070/files#r1595622740